### PR TITLE
 Fix for a jit liveness bug.

### LINKED ISSUE #24253: Register allocator assert: "Expected empty defList at end of block"
This test below causes an assert on x64 Windows (and likely elsewhere).
The assert is 

```
Assert failure(PID 18440 [0x00004808], Thread: 26224 [0x6670]): Assertion failed '!"Expected empty defList at end of block"' in 'GitHub_999999999.Program:GetStruct(int):struct' (IL size 24)

    File: f:\coreclr7\src\jit\lsrabuild.cpp Line: 2210
    Image: F:\coreclr7\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe
```

```
using System;
using System.Runtime.CompilerServices;

namespace N
{
    struct TestStruct
    {
        public static readonly TestStruct ZeroStruct = new TestStruct(0);

        [MethodImpl(MethodImplOptions.NoInlining)]
        public TestStruct(int i)
        {
            this.i = i;
            this.j = 5;
            this.k = 5;
            this.l = 5;
        }

        int i;
        int j;
        short k;
        short l;
    }

    class Program
    {
        static int Main(string[] args)
        {
            GetStruct(1);
            return 100;
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        private static TestStruct GetStruct(int i)
        {
            TestStruct result = TestStruct.ZeroStruct;
            try
            {
                result = new TestStruct(i);
            }
            catch
            {
                throw new ArgumentException();
            }
            return result;
        }
    }
}

```

This test is extracted from DevDiv853438.

### PR DESCRIPTION
**Pull request by erozenfeld**
`fgRemoveDeadStore` has special logic for removing dead assignments
whose rhs was of type `TYP_STRUCT`:

https://github.com/dotnet/coreclr/blob/311b5e2fe413c6c74a2a3680ab54d8a978651472/src/jit/liveness.cpp#L2264-L2274

That logic was applied to "normal"
assignments (i.e., direct children of `GT_STMT`) but not to "internal"
assignments (e.g., children of `GT_COMMA`).
The test case has an internal assignment and, because this logic wasn't
applied, we ended up with a standalone `GT_IND` of type `TYP_STRUCT`
that the register allocator can't handle.

This change apples the missing logic to "internal" assignments.

The second commit cleans up `fgRemoveDeadStore`.

Fixes #24253.

### CODE REVIEW

**Comment by erozenfeld** [comment]
No x64 pmi diffs in frameworks and benchmarks.
I will try to check SPMI diffs as well.

**Comment by erozenfeld** [comment]
@dotnet/jit-contrib @sandreenko @CarolEidt PTAL
It's probably easier to review the two commits individually: the first one is the actual fix and the second one is the cleanup of confusing surrounding code.

File: src/jit/liveness.cpp [hunk 1]
```diff
@@ -2245,98 +2272,49 @@ bool Compiler::fgRemoveDeadStore(GenTree**        pTree,
...
-                    {
-                        printf("Extracted side effects list...\n");
-                        gtDispTree(sideEffList);
-                        printf("\n");
-                    }
-#endif // DEBUG
 
-                    /* Replace the assignment statement with the list of side effects */
-                    noway_assert(sideEffList->gtOper != GT_STMT);
+                /* Replace the assignment statement with the list of side effects */
```
Comment (CarolEidt) [inline src/jit/liveness.cpp]: While you're here you could make these C++ style comments.

File: src/jit/liveness.cpp [hunk 2]
```diff
@@ -2409,7 +2369,6 @@ bool Compiler::fgRemoveDeadStore(GenTree**        pTree,
             }
             else
             {
-            NO_SIDE_EFFECTS:
```
Comment (CarolEidt) [inline src/jit/liveness.cpp]: I love getting rid of `goto`s!!

File: src/jit/liveness.cpp [hunk 3]
```diff
@@ -2359,6 +2359,19 @@ bool Compiler::fgRemoveDeadStore(GenTree**        pTree,
...
 #endif // DEBUG
+
+                if (rhsNode->TypeGet() == TYP_STRUCT)
+                {
+                    // This is a block assignment. An indirection of the rhs is not considered to
+                    // happen until the assignment, so we will extract the side effects from only
+                    // the address.
+                    if (rhsNode->OperIsIndir())
+                    {
+                        assert(rhsNode->OperGet() != GT_NULLCHECK);
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: Why should not it be a nullcheck?

File: src/jit/liveness.cpp [hunk 4]
```diff
@@ -2245,98 +2272,49 @@ bool Compiler::fgRemoveDeadStore(GenTree**        pTree,
             noway_assert(compCurStmt->gtStmtExpr == asgNode);
             JITDUMP("top level assign\n");
 
-            /* Check for side effects */
-
-            if (rhsNode->gtFlags & GTF_SIDE_EFFECT)
+            if (sideEffList)
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: ` if (sideEffList != nullptr)`

File: src/jit/liveness.cpp [hunk 5]
```diff
@@ -2245,98 +2272,49 @@ bool Compiler::fgRemoveDeadStore(GenTree**        pTree,
...
-                    }
-#endif // DEBUG
 
-                    /* Replace the assignment statement with the list of side effects */
-                    noway_assert(sideEffList->gtOper != GT_STMT);
+                /* Replace the assignment statement with the list of side effects */
+                noway_assert(sideEffList->gtOper != GT_STMT);
 
-                    *pTree = compCurStmt->gtStmtExpr = sideEffList;
+                *pTree = compCurStmt->gtStmtExpr = sideEffList;
```
Comment (sandreenko) [inline src/jit/liveness.cpp]: Is it possible that `pTree != &compCurStmt->gtStmtExpr`?

File: src/jit/liveness.cpp (hunk 3 above)
Comment (erozenfeld) [reply src/jit/liveness.cpp]: We never create GT_NULLCHECK as a rhs of an assignment. I guess this assert just makes sure we don't accidentally remove a null check.

File: src/jit/liveness.cpp (hunk 5 above)
Comment (erozenfeld) [reply src/jit/liveness.cpp]: pTree is an address of a local GenTree* variable in the caller, (e.g., address of `tree` in `fgComputeLife`.

File: src/jit/liveness.cpp [hunk 6]
```diff
@@ -2235,6 +2235,33 @@ bool Compiler::fgRemoveDeadStore(GenTree**        pTree,
             return false;
         }
 
+        // Check for side effects
+        GenTree* sideEffList = nullptr;
+        if (rhsNode->gtFlags & GTF_SIDE_EFFECT)
+        {
+// Extract the side effects
```
[the remark this answers is not in the document]
Comment (erozenfeld) [reply src/jit/liveness.cpp]: Done.

File: src/jit/liveness.cpp (hunk 1 above)
Comment (erozenfeld) [reply src/jit/liveness.cpp]: Done.

File: src/jit/liveness.cpp (hunk 4 above)
Comment (erozenfeld) [reply src/jit/liveness.cpp]: Done.

**Comment by erozenfeld** [comment]
No desktop SPMI diffs.

**Comment by erozenfeld** [comment]
CoreFX test failed with
System.Net.Security.Tests.SslStreamCredentialCacheTest.SslStream_SameCertUsedForClientAndServer_Ok [FAIL]
      System.TimeoutException : Task timed out after 15000ms

**Comment by erozenfeld** [comment]
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests

**Comment by erozenfeld** [comment]
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests

**Comment by erozenfeld** [comment]
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests (the broken machine is disabled now).

**Pull request merged**